File: backend/app/models/transaction.py
"""Pydantic schemas for transactions."""

from datetime import datetime
from decimal import Decimal
from pydantic import BaseModel, Field


class TransactionResponse(BaseModel):
    """Single transaction in API response."""
    id: str
    merchant_name: str
    category: str
    amount: Decimal
    currency: str
    payment_method: str
    status: str
    txn_date: datetime
    created_at: datetime


class PaginationMeta(BaseModel):
    """Pagination metadata."""
    page: int
    limit: int
    total: int
    total_pages: int


class TransactionListResponse(BaseModel):
    """Paginated list of transactions."""
    data: list[TransactionResponse]
    pagination: PaginationMeta


class TransactionFilters(BaseModel):
    """Query parameters for filtering transactions."""
    page: int = Field(default=1, ge=1)
    limit: int = Field(default=20, ge=1, le=100)
    category: str | None = None
    status: str | None = None
    date_from: str | None = None
    date_to: str | None = None
    amount_min: float | None = None
    amount_max: float | None = None
    search: str | None = None
    sort_by: str = Field(default="txn_date")
    sort_dir: str = Field(default="desc")

File: frontend/types/transaction.ts
/**
 * TypeScript interfaces matching backend Pydantic models.
 */

export interface Transaction {
  id: string;
  merchant_name: string;
  category: string;
  amount: number;
  currency: string;
  payment_method: string;
  status: 'success' | 'failed' | 'pending';
  txn_date: string;
  created_at: string;
}

export interface PaginationMeta {
  page: number;
  limit: number;
  total: number;
  total_pages: number;
}

export interface TransactionListResponse {
  data: Transaction[];
  pagination: PaginationMeta;
}

export interface TransactionFilters {
  page?: number;
  limit?: number;
  category?: string;
  status?: string;
  date_from?: string;
  date_to?: string;
  amount_min?: number;
  amount_max?: number;
  search?: string;
  sort_by?: string;
  sort_dir?: 'asc' | 'desc';
}

export interface WalletResponse {
  id: number;
  coin_balance: number;
}

export interface RewardItem {
  id: number;
  name: string;
  description: string | null;
  coin_cost: number;
}

export interface RewardsListResponse {
  data: RewardItem[];
}

export interface RedemptionResponse {
  id: string;
  reward_id: number;
  reward_name: string;
  coins_spent: number;
  status: 'confirmed' | 'failed';
  new_balance: number;
  created_at: string;
}

export interface CategoryBreakdown {
  category: string;
  total_amount: number;
  transaction_count: number;
}

export interface MonthlyTrend {
  month: string;
  total_amount: number;
  transaction_count: number;
}

export interface SummaryStats {
  total_transactions: number;
  total_spend: number;
  success_count: number;
  failed_count: number;
  pending_count: number;
}
